File: population-service/src/metrics_interceptor.py
import datetime
from typing import Callable
import grpc
from grpc_interceptor import ServerInterceptor
from prometheus import Prometheus, CallMetrics


class MetricsInterceptor(ServerInterceptor):
    def __init__(self, metrics_provider: Prometheus):
        self.metrics_provider = metrics_provider

    def intercept(
        self,
        method: Callable,
        request,
        context: grpc.ServicerContext,
        method_name: str,
    ):
        request_start = datetime.datetime.now()

        try:
            result_status = grpc.StatusCode.OK
            return method(request, context)
        except grpc.RpcError as e:
            result_status = e.__dict__['code']
        except Exception:
            result_status = grpc.StatusCode.UNKNOWN
        finally:
            metrics = CallMetrics(method_name, 'unary', result_status, request_start)
            self.metrics_provider.collect_call_metrics(metrics)

File: lib/Utils.Grpc/GrpcClients/Interceptors/MetricsInterceptor.cs
using Utils.Grpc.Metrics;
using Grpc.Core;
using Grpc.Core.Interceptors;
using System;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Utils.Grpc.GrpcClients.Interceptors
{
    public class MetricsInterceptor : Interceptor
    {
        private readonly IMetricsProvider _metricsProvider;

        public MetricsInterceptor(IMetricsProvider metricsProvider)
        {
            _metricsProvider = metricsProvider;
        }

        public override AsyncUnaryCall<TResponse> AsyncUnaryCall<TRequest, TResponse>(TRequest request,
                                                                                      ClientInterceptorContext<TRequest, TResponse> context,
                                                                                      AsyncUnaryCallContinuation<TRequest, TResponse> continuation)
        {
            var requestStart = DateTime.UtcNow;
            var call = continuation(request, context);
            return new AsyncUnaryCall<TResponse>(
                call.ResponseAsync.ContinueWith(task =>
                {
                    Handled(MethodType.Unary, context, call.GetStatus().StatusCode, requestStart).Wait();
                    return task.Result;
                }),
                call.ResponseHeadersAsync, call.GetStatus, call.GetTrailers, call.Dispose);
        }

        
        private Task Handled<TRequest, TResponse>(MethodType type, ClientInterceptorContext<TRequest, TResponse> context, StatusCode status, DateTime requestStart)
            where TRequest : class
            where TResponse : class
            =>  _metricsProvider.CollectCallMetrics(new CallData {
                CallType = type.ToString().ToLower(),
                Duration = DateTime.UtcNow.Subtract(requestStart).TotalMilliseconds,
                Method = GetMethod(context),
                Status = status.ToString().ToLower()
            });

        private string GetMethod<TRequest, TResponse>(ClientInterceptorContext<TRequest, TResponse> context)
            where TRequest : class
            where TResponse : class
        {
            var target = context.Options.Headers.FirstOrDefault(i => i.Key == "target").Value;
            var method = new StringBuilder();
            if (!string.IsNullOrEmpty(target)) method.Append($"{target}-");
            method.Append(context.Method.Name);

            return method.ToString();
        }
    }
} 
